### ISSUE TITLE
Anupq 3.1.4 fails to build with new buildsystem

### ISSUE THREAD

**Issue by hungaborhorvath** [gap-packages/anupq#20]
I am not sure if this was reported or not. Anupq 3.1.4 builds properly with ````bin/BuildPackages.sh```` for GAP stable 4.8.7 on my sysem, but fails to build with current master. I am using debian jessie 64 bit. Let me know if you ant to know if any packages are installed.

I am not sure what could be the culprit, but here is ````log/anupq-3.1.4.log````: 

````
2017. ápr. 15., szombat, 23.01.29 CEST

==== Checking anupq-3.1.4
Running './configure' 'CFLAGS=-m32' 'LDFLAGS=-m32' '--with-gaproot=/home/ghorvath/work/gap' 
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking for gawk... (cached) mawk
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/home/ghorvath/work/gap/pkg/anupq-3.1.4':
configure: error: C compiler cannot create executables
See `config.log' for more details

WARNING: Failed to build anupq-3.1.4
````
The ````config.log```` that is cited above is this: 
````
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by ANU p-Quotient Program configure 3.1.3, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  $ ./configure CFLAGS=-m32 LDFLAGS=-m32 --with-gaproot=/home/ghorvath/work/gap

## --------- ##
## Platform. ##
## --------- ##

hostname = SamsungLaptop
uname -m = x86_64
uname -r = 4.9.0-0.bpo.2-amd64
uname -s = Linux
uname -v = #1 SMP Debian 4.9.13-1~bpo8+1 (2017-02-27)

/usr/bin/uname -p = unknown
/bin/uname -X     = unknown

/bin/arch              = unknown
/usr/bin/arch -k       = unknown
/usr/convex/getsysinfo = unknown
/usr/bin/hostinfo      = unknown
/bin/machine           = unknown
/usr/bin/oslevel       = unknown
/bin/universe          = unknown

PATH: /home/ghorvath/bin
PATH: /usr/sbin
PATH: /sbin
PATH: /usr/local/bin
PATH: /usr/bin
PATH: /bin
PATH: /usr/local/games
PATH: /usr/games


## ----------- ##
## Core tests. ##
## ----------- ##

configure:2275: checking for a BSD-compatible install
configure:2343: result: /usr/bin/install -c
configure:2354: checking whether build environment is sane
configure:2409: result: yes
configure:2560: checking for a thread-safe mkdir -p
configure:2599: result: /bin/mkdir -p
configure:2606: checking for gawk
configure:2636: result: no
configure:2606: checking for mawk
configure:2622: found /usr/bin/mawk
configure:2633: result: mawk
configure:2644: checking whether make sets $(MAKE)
configure:2666: result: yes
configure:2695: checking whether make supports nested variables
configure:2712: result: yes
configure:2849: checking whether make supports nested variables
configure:2866: result: yes
configure:2889: checking for gawk
configure:2916: result: mawk
configure:2975: checking for gcc
configure:2991: found /usr/bin/gcc
configure:3002: result: gcc
configure:3231: checking for C compiler version
configure:3240: gcc --version >&5
gcc (Debian 4.9.2-10) 4.9.2
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

configure:3251: $? = 0
configure:3240: gcc -v >&5
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/4.9/lto-wrapper
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 4.9.2-10' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --with-arch-32=i586 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.9.2 (Debian 4.9.2-10) 
configure:3251: $? = 0
configure:3240: gcc -V >&5
gcc: error: unrecognized command line option '-V'
gcc: fatal error: no input files
compilation terminated.
configure:3251: $? = 4
configure:3240: gcc -qversion >&5
gcc: error: unrecognized command line option '-qversion'
gcc: fatal error: no input files
compilation terminated.
configure:3251: $? = 4
configure:3271: checking whether the C compiler works
configure:3293: gcc -m32  -m32 conftest.c  >&5
/usr/bin/ld: cannot find crt1.o: No such file or directory
/usr/bin/ld: cannot find crti.o: No such file or directory
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/4.9/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/4.9/libgcc_s.so when searching for -lgcc_s
/usr/bin/ld: cannot find -lgcc_s
/usr/bin/ld: cannot find -lc
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/4.9/libgcc.a when searching for -lgcc
/usr/bin/ld: cannot find -lgcc
/usr/bin/ld: skipping incompatible /usr/lib/gcc/x86_64-linux-gnu/4.9/libgcc_s.so when searching for -lgcc_s
/usr/bin/ld: cannot find -lgcc_s
/usr/bin/ld: cannot find crtn.o: No such file or directory
collect2: error: ld returned 1 exit status
configure:3297: $? = 1
configure:3335: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "ANU p-Quotient Program"
| #define PACKAGE_TARNAME "anupq"
| #define PACKAGE_VERSION "3.1.3"
| #define PACKAGE_STRING "ANU p-Quotient Program 3.1.3"
| #define PACKAGE_BUGREPORT "https://github.com/gap-packages/anupq/issues"
| #define PACKAGE_URL "http://gap-packages.github.io/anupq/"
| #define PACKAGE "anupq"
| #define VERSION "3.1.3"
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:3340: error: in `/home/ghorvath/work/gap/pkg/anupq-3.1.4':
configure:3342: error: C compiler cannot create executables
See `config.log' for more details

## ---------------- ##
## Cache variables. ##
## ---------------- ##

ac_cv_env_CC_set=
ac_cv_env_CC_value=
ac_cv_env_CFLAGS_set=set
ac_cv_env_CFLAGS_value=-m32
ac_cv_env_CONFIGNAME_set=
ac_cv_env_CONFIGNAME_value=
ac_cv_env_CPPFLAGS_set=
ac_cv_env_CPPFLAGS_value=
ac_cv_env_CPP_set=
ac_cv_env_CPP_value=
ac_cv_env_LDFLAGS_set=set
ac_cv_env_LDFLAGS_value=-m32
ac_cv_env_LIBS_set=
ac_cv_env_LIBS_value=
ac_cv_env_build_alias_set=
ac_cv_env_build_alias_value=
ac_cv_env_host_alias_set=
ac_cv_env_host_alias_value=
ac_cv_env_target_alias_set=
ac_cv_env_target_alias_value=
ac_cv_path_install='/usr/bin/install -c'
ac_cv_path_mkdir=/bin/mkdir
ac_cv_prog_AWK=mawk
ac_cv_prog_ac_ct_CC=gcc
ac_cv_prog_make_make_set=yes
am_cv_make_support_nested_variables=yes

## ----------------- ##
## Output variables. ##
## ----------------- ##

ACLOCAL='${SHELL} /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing aclocal-1.15'
AMDEPBACKSLASH=''
AMDEP_FALSE=''
AMDEP_TRUE=''
AMTAR='$${TAR-tar}'
AM_BACKSLASH='\'
AM_DEFAULT_V='$(AM_DEFAULT_VERBOSITY)'
AM_DEFAULT_VERBOSITY='0'
AM_V='$(V)'
AUTOCONF='${SHELL} /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing autoconf'
AUTOHEADER='${SHELL} /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing autoheader'
AUTOMAKE='${SHELL} /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing automake-1.15'
AWK='mawk'
CC='gcc'
CCDEPMODE=''
CFLAGS='-m32'
CONFIGNAME=''
CPP=''
CPPFLAGS=''
CYGPATH_W='echo'
DEFS=''
DEPDIR=''
ECHO_C=''
ECHO_N='-n'
ECHO_T=''
EGREP=''
EXEEXT=''
GAPARCH=''
GAPROOT=''
GAP_CPPFLAGS=''
GAP_EXEC=''
GMP_CPPFLAGS=''
GMP_LIBS=''
GREP=''
INSTALL_DATA='${INSTALL} -m 644'
INSTALL_PROGRAM='${INSTALL}'
INSTALL_SCRIPT='${INSTALL}'
INSTALL_STRIP_PROGRAM='$(install_sh) -c -s'
LDFLAGS='-m32'
LIBOBJS=''
LIBS=''
LTLIBOBJS=''
MAKEINFO='${SHELL} /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing makeinfo'
MKDIR_P='/bin/mkdir -p'
OBJEXT=''
PACKAGE='anupq'
PACKAGE_BUGREPORT='https://github.com/gap-packages/anupq/issues'
PACKAGE_NAME='ANU p-Quotient Program'
PACKAGE_STRING='ANU p-Quotient Program 3.1.3'
PACKAGE_TARNAME='anupq'
PACKAGE_URL='http://gap-packages.github.io/anupq/'
PACKAGE_VERSION='3.1.3'
PATH_SEPARATOR=':'
SED=''
SET_MAKE=''
SHELL='/bin/bash'
STRIP=''
VERSION='3.1.3'
ac_ct_CC='gcc'
am__EXEEXT_FALSE=''
am__EXEEXT_TRUE=''
am__fastdepCC_FALSE=''
am__fastdepCC_TRUE=''
am__include=''
am__isrc=''
am__leading_dot='.'
am__nodep=''
am__quote=''
am__tar='$${TAR-tar} chof - "$$tardir"'
am__untar='$${TAR-tar} xf -'
bindir='${exec_prefix}/bin'
build_alias=''
datadir='${datarootdir}'
datarootdir='${prefix}/share'
docdir='${datarootdir}/doc/${PACKAGE_TARNAME}'
dvidir='${docdir}'
exec_prefix='NONE'
host_alias=''
htmldir='${docdir}'
includedir='${prefix}/include'
infodir='${datarootdir}/info'
install_sh='${SHELL} /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/install-sh'
libdir='${exec_prefix}/lib'
libexecdir='${exec_prefix}/libexec'
localedir='${datarootdir}/locale'
localstatedir='${prefix}/var'
mandir='${datarootdir}/man'
mkdir_p='$(MKDIR_P)'
oldincludedir='/usr/include'
pdfdir='${docdir}'
prefix='NONE'
program_transform_name='s,x,x,'
psdir='${docdir}'
sbindir='${exec_prefix}/sbin'
sharedstatedir='${prefix}/com'
sysconfdir='${prefix}/etc'
target_alias=''

## ----------- ##
## confdefs.h. ##
## ----------- ##

/* confdefs.h */
#define PACKAGE_NAME "ANU p-Quotient Program"
#define PACKAGE_TARNAME "anupq"
#define PACKAGE_VERSION "3.1.3"
#define PACKAGE_STRING "ANU p-Quotient Program 3.1.3"
#define PACKAGE_BUGREPORT "https://github.com/gap-packages/anupq/issues"
#define PACKAGE_URL "http://gap-packages.github.io/anupq/"
#define PACKAGE "anupq"
#define VERSION "3.1.3"

configure: exit 77
````

**Comment by fingolfin** [comment]
Actually, for me it does build just fine.

Your error suggests to me that you don't have a 32bit compiler installed. We now require that for anupq, as compiling it as 64bit produces a broken binary.

If you are on debian or ubuntu, try install `gcc-multilib` and see if that helps. Ah, you'll also need `libgmp-dev:i386`

**Comment by hungaborhorvath** [comment]
@fingolfin Thanks, I will check whether that works, and get back to you. 

However, I did not see anywhere that a 32 bit compiler is needed, not even in the README file. It might be a good idea to suggest the necessary packages on this  webpage, as well: 
https://www.gap-system.org/Download/tools.html

**Comment by hungaborhorvath** [comment]
@fingolfin Nope, still no luck, after installing both ````gcc-multilib```` and ````libgmp-dev:i386````, I get the following (new) error:
````
ghorvath@HorvathGabor-PC:~/work/gap/pkg$ ../bin/BuildPackages.sh anupq-3.1.4
Using GAP root /home/ghorvath/work/gap
Building with 64-bit ABI
Attempting to build GAP packages.
Note that many GAP packages require extra programs to be installed,
and some are quite difficult to build. Please read the documentation for
packages which fail to build correctly, and only worry about packages
you require!

2017. júl. 20., csütörtök, 16.20.07 CEST

==== Checking anupq-3.1.4
Running './configure' 'CFLAGS=-m32' 'LDFLAGS=-m32' '--with-gaproot=/home/ghorvath/work/gap' 
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether make supports nested variables... (cached) yes
checking for gawk... (cached) mawk
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... gcc3
checking whether make sets $(MAKE)... (cached) yes
checking for a sed that does not truncate output... /bin/sed
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for stdlib.h... (cached) yes
checking for string.h... (cached) yes
checking sys/time.h usability... yes
checking sys/time.h presence... yes
checking for sys/time.h... yes
checking for CONFIGNAME... none
checking for GAP root directory... /home/ghorvath/work/gap
checking for GAP architecture... x86_64-pc-linux-gnu-gcc-default64
checking for GAP include files... /home/ghorvath/work/gap/src/compiled.h
checking for GAP config.h... /home/ghorvath/work/gap/bin/x86_64-pc-linux-gnu-gcc-default64/config.h
checking for GAP's gmp.h location... /home/ghorvath/work/gap/bin/x86_64-pc-linux-gnu-gcc-default64/extern/gmp/include/gmp.h
checking gmp.h usability... yes
checking gmp.h presence... yes
checking for gmp.h... yes
checking whether linking against GMP works... no
checking for stdlib.h... (cached) yes
checking for GNU libc compatible malloc... yes
checking for stdlib.h... (cached) yes
checking for GNU libc compatible realloc... yes
checking for pid_t... yes
checking vfork.h usability... no
checking vfork.h presence... no
checking for vfork.h... no
checking for fork... yes
checking for vfork... yes
checking for working fork... yes
checking for working vfork... (cached) yes
checking for sys/wait.h that is POSIX.1 compatible... yes
checking for gethostname... yes
checking for waitpid... yes
checking for wait4... yes
checking for tmpnam... yes
checking for tempnam... yes
checking for strftime... yes
checking for library containing log10... -lm
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating testPq
config.status: creating include/config.h
config.status: include/config.h is unchanged
config.status: executing depfiles commands
Running 'make' 'CFLAGS=-m32' 'LOPTS=-m32' 
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/bash /home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing aclocal-1.15 -I m4
/home/ghorvath/work/gap/pkg/anupq-3.1.4/cnf/missing: sor: 81: aclocal-1.15: parancs nem található
WARNING: 'aclocal-1.15' is missing on your system.
You should only need it if you modified 'acinclude.m4' or
'configure.ac' or m4 files included by 'configure.ac'.
The 'aclocal' program is part of the GNU Automake package:
<http://www.gnu.org/software/automake>
It also requires GNU Autoconf, GNU m4 and Perl in order to run:
<http://www.gnu.org/software/autoconf>
<http://www.gnu.org/software/m4/>
<http://www.perl.org/>
Makefile:567: recipe for target 'aclocal.m4' failed
make: *** [aclocal.m4] Error 127

WARNING: Failed to build anupq-3.1.4

Packages which failed to build are in ./log/fail.log
ghorvath@HorvathGabor-PC:~/work/gap/pkg$ 
````
Do I need to give some 32-bit parameters to ````BuildPackages.sh````?

**Comment by fingolfin** [comment]
First off, the error you get seems to be caused by a bug in how the GAP 4.8.7 distribution was made; some file modifications dates are wrong. I also (by coincidence, for completely unrelated reasons) noticed this today.
@alex-konovalov I thought the new distribution scripts took care to set equal timestamps of all files? Or was that not yet used when GAP 4.8.7 was created?

Anyway, as a workaround, try this inside the `anupq` directory:
```sh
touch configure Makefile.in aclocal.m4
```
Then try again to build it.

Finally, the version of `BuildPackages.sh` in the `master` branch of GAP gives all the right flags for 32 bit, and works fine for me on other systems (see also e.g. https://github.com/gap-system/gap/pull/1512 where this has been running multiple times today)

**Comment by hungaborhorvath** [comment]
Indeed, it compiles now. So all is left to do maybe to check in ````BuildPackages.sh```` for whether these 3 files exist, and let people know on the website about the extra packages to be installed.

**Comment by fingolfin** [comment]
Not sure which files you want to check for... These files all *do* exist, but due to a bug (IMHO -- people may have different opinion on this :-) in how the GAP package distribution was created, their timestamps are "wrong", causing the build system to think they need to be regenerated.
In other words: This should be fixed in `BuildPackages.sh`.

I have serious reservations about editing that webpage you linked. It is super incomplete anyway, hardly anybody knows about it, and the whole website is for the most part unmaintained. It doesn't scale well to add information on individual packages there, either. Instead, we could (a) enhance `BuildPackages.sh` to detect whether a 32bit compiler is available, and if not, issue a more helpful error when trying to use it; and (b) of course enhance the documentation of the anupq package to mention the troubles on 64bit, and the strong recommendation to use a 32bit compiler. I might even integrate code to check for a 32bit compiler, and enforce it. We'll see.

**Comment by alex-konovalov** [comment]
The new distribution scripts took care to set equal timestamps of all files, indeed - in the archive which has GAP core and packages, they should be equal. I am not sure if this also happens if one of `bootstrap-pkg` targets has been used (but there should be no need to use it while installing GAP 4.8.7).

**Comment by hungaborhorvath** [comment]
@fingolfin Ok, so the issue with the files needs to be fixed in ````BuildPackages.sh````, unless it is already fixed as @alex-konovalov says.

That webpage is less outdated than you think. For the package dependencies, I have checked them about half year ago, in fact even did a PR on it, so the information is now almost up-to-date. The packages that do not compile for me (using the instructions on that webpage) are 
````
anupq
NormalizInterface-0.9.8
pargap
PolymakeInterface
linboxing
float-0.7.5
Browse
````
For ````Browse```` I have emailed Thomas Breuer some time ago, so he is aware of the issue. I did not care to check the others, yet, because I do not need them. So apart from these packages, the information seems accurate, and I even added there the debian/ubuntu package names.
